### ISSUE TITLE
PEtab SciML support: hybrid ODE and neural network problems

### ISSUE THREAD

**Issue by matthiaskoenig** [matthiaskoenig/sbmlsim#207]
## Summary

Support [PEtab SciML](https://github.com/PEtab-dev/petab_sciml), the PEtab v2 extension for parameter estimation problems in which a mechanistic ODE model (SBML) is hybridized with neural networks (NN) and the mechanistic and the NN parameters are estimated jointly from time series data. `sbmlsim` already reads and writes PEtab v2 problems (`sbmlsim.fit.petab_v2`, #202), so the SciML extension is the next layer on top of the reader and the exporter: the extension block of the problem YAML, the NN model YAML, the HDF5 array files, the hybridization table and the NN entries of the mapping and the parameter table. The goal is that a problem of the [PEtab SciML test suite](https://github.com/PEtab-dev/petab_sciml_testsuite) is read into an `OptimizationProblem`, simulated with roadrunner, evaluated and fitted, and that the results match the reference values of the test suite.

Reference: Persson S, Snelling B, Philipps M, Weindl D, Cvijovic M, Hasenauer J, Pathirana D, Fröhlich F. *PEtab SciML: an exchange format for specifying and training dynamic scientific machine learning models.* arXiv:2608.20184 (2026), https://arxiv.org/abs/2608.20184

### What PEtab SciML is

PEtab SciML keeps the core PEtab v2 tables (models, conditions, experiments, observables, measurements, parameters, mapping), so every PEtab v2 feature (multiple conditions and experiments, events and dosing, pre-equilibration, priors, noise models) applies to a hybrid problem, and adds three elements:

1. **The NN model.** A language independent YAML format (`format: YAML`) which lists the `layers` (PyTorch names and keyword arguments, e.g. `Linear(in_features, out_features, bias)`, `Conv2d`, `MaxPool2d`, `Flatten`, `Dropout`, ...) and the `forward` pass as a list of ops (`placeholder`, `call_module`, `call_method`/`call_function` for activations and tensor ops, `output`), i.e. the traced PyTorch FX graph. The `petab_sciml` python package exports a `torch.nn.Module` to this format. A problem may also ship a framework native model (`format: pytorch | equinox | lux.jl`), which an importer is free to not support.
2. **The hybridization.** A `pre_initialization` flag per network in the extension block plus the hybridization table (`targetId`, `targetValue`) and the NN entries of the mapping table (`petabEntityId` → `netId.inputs[i][j]`, `netId.outputs[i][j]`, `netId.parameters`, `netId.parameters[layerId].weight`) say where the network sits:
   - **pre-initialization** (`pre_initialization: true`): the inputs are constants (parameters of the parameter table, of a condition, or an `array` from an HDF5 file), the outputs set non-estimated parameters or initial values before the simulation. Evaluated once per condition.
   - **ODE right hand side** (`pre_initialization: false`, output targets are parameters of the ODE): inputs are expressions of species, parameters and time, the outputs replace parameters of the rate equations. Evaluated at every step of the integrator.
   - **observable** (`pre_initialization: false`, output referenced in an `observableFormula`): the network maps the simulated state to the measurement.
   
   A problem may combine several networks and several patterns.
3. **Array data.** One HDF5 file per network (or shared) with the groups `metadata` (`pytorch_format`), `inputs/<inputId>/<conditionIds or 0>` and `parameters/<netId>/<layerId>/<weight|bias>`, row major with PyTorch indexing. The parameter table refers to the arrays with `nominalValue = array`; `estimate`, `lowerBound`/`upperBound` (`-inf`/`inf` or empty) and the priors apply to the whole network, to a layer (`net1.parameters[layer1]`) or to one array (`net1.parameters[layer1].weight`), the more specific row wins, so single layers or arrays are frozen.

Example (test case 001, Lotka-Volterra with a network for one interaction term, `dpredator/dt = N(prey, predator) - delta * predator` where the SBML model has `gamma` in place of `N`):

```yaml
# problem.yaml
extensions:
  sciml:
    version: "0.1.0"
    required: true
    neural_networks:
      net1: {location: "net1.yaml", format: "YAML", pre_initialization: false}
    hybridization_files: ["hybridization.tsv"]
    array_files: ["net1_ps.hdf5"]
```

```
# hybridization.tsv            # mapping.tsv
targetId     targetValue       petabEntityId   modelEntityId
net1_input1  prey              net1_input1     net1.inputs[0][0]
net1_input2  predator          net1_input2     net1.inputs[0][1]
gamma        net1_output1      net1_output1    net1.outputs[0][0]
                               net1_ps         net1.parameters
# parameters.tsv
parameterId  parameterScale  lowerBound  upperBound  nominalValue  estimate
alpha        lin             0.0         15.0        1.3           true
net1_ps      lin             -inf        inf         array         true
```

Specification: https://petab-sciml.readthedocs.io/latest/format.html, supported layers and activations: https://petab-sciml.readthedocs.io/latest/layers.html, reference library: https://pypi.org/project/petab-sciml/ (`h5py`, `numpy`, `pydantic`, `ruamel.yaml`, optional `torch`).

### Existing implementations

- **PEtab.jl** (Julia, Lux.jl): `ext/PEtabLuxExtension/` (`parse.jl`, `ml_model.jl`, `read_ps.jl`, `write_ps.jl`) parses the NN YAML into a Lux chain, `ext/PEtabSciMLSensitivityExtension/` adds adjoint gradients, `test/petab_sciml_testsuite/` runs the test suite, one file per case. Tutorials: https://github.com/sebapersson/PEtab.jl/tree/main/docs/src/tutorials/sciml. Supports every layer of the format.
- **AMICI** (Python/JAX): `python/sdist/amici/jax/petab.py` imports a problem into equinox modules and diffrax solvers, `jax/model.py`, `jax/ode_export.py`; gradients by automatic differentiation of the whole solve. Supports a subset of the layers.
- **petab_sciml** (Python reference library): `standard/nn_model.py` (pydantic model of the NN YAML, PyTorch import/export), `standard/array_data.py` (HDF5 arrays), `problem_utils/neural_ode.py` (helper to build a neural ODE problem), linting, transformations for multiple shooting and curriculum learning.

### The test suite

https://github.com/PEtab-dev/petab_sciml_testsuite, `test_cases/` in three groups, all files pre-built:

- `ml_model_import/001`–`054`: only the network. `net.yaml`, three inputs, parameter sets and expected outputs (`net_input_i.hdf5`, `net_ps_i.hdf5`, `net_output_i.hdf5`), `solutions.yaml` with the combinations to check and the axis orders (`input_order_py`, ...); dropout cases give the number of forward passes to average. Covers `Conv1-3d`, `ConvTranspose1-3d`, the pooling layers, `BatchNorm`, `InstanceNorm`, `LayerNorm`, the dropouts, `Linear`, `Bilinear`, `Flatten` and all activations.
- `sciml_problem_import/001`–`039`: the hybrid problems, all with feed-forward `Linear` networks and `tanh`, on the Lotka-Volterra model. `petab/` holds the problem, `solutions.yaml` the negative log-likelihood `llh` with `tol_llh`, the simulations at the measurement points (`simulations.tsv`, `tol_simulations`) and the gradient of the objective at the nominal parameters, split into `grad_mech.tsv` for the mechanistic and `grad_<net>.hdf5` for the NN parameters (`tol_grad`). The cases cover: NN in the RHS (001, 010, 016, 019, 023, 026, 029, 035, 036, 038), NN setting a parameter or an initial value before the simulation with inputs from the parameter table, a condition, an estimated parameter or an array file (002, 003, 005–008, 014, 015, 017, 021, 024, 027, 030, 032–034), NN in an observable (004, 009, 018, 020, 025, 028, 031, 037, 039), mixtures (011–013, 022), frozen networks, layers and arrays (016–018, 023–028), two outputs (019–022), two input arguments (029), a math expression as input (035), priors on the NN parameters (032–034).
- `initialization/001`–`003`: the nominal NN parameters after the import when layers are frozen, `solutions.yaml` names the reference `net1_ref.hdf5`.

## Design for sbmlsim

`sbmlsim` simulates with libroadrunner, which integrates SBML and nothing else, and its objective is `0.5 * Σ r²` of the weighted residuals with scipy least squares or differential evolution over the parameters of the parameter table (`FitParameter`). Neither is changed by this issue; the three patterns are mapped onto what exists:

1. **Pre-initialization network → changes.** The network is evaluated in numpy once per condition/experiment with the inputs of the parameter table, the condition or the array file, and its outputs become `changes` of the `Timecourse` (parameter or initial value), i.e. the same mechanism a condition of the PEtab v2 reader uses. When an input is an estimated parameter (005, 006) the evaluation is part of the objective, so it moves into `OptimizationProblem.residuals` before the simulation.
2. **Observable network → post processing.** The network is evaluated in numpy on the simulated trajectories, the same place a formula observable is evaluated (`Data` of type `FUNCTION`, `petab_v2/observables.py`).
3. **RHS network → SBML.** The network is compiled into the SBML model: one parameter per hidden unit and output with an assignment rule, i.e. `h1_j = tanh(Σ_i W_ji * x_i + b_j)` written as MathML, and the target parameter (`gamma`) gets the assignment rule of the output; every weight and bias becomes an SBML parameter, so the fit sets them through the usual roadrunner selections and the integrator sees a plain ODE system. This is the same approach as `petab_v2/observables.py`, which writes the observables of a problem into the model as assignment rules. `Linear`, `Bilinear` and `Flatten` and all elementwise activations are expressible in MathML (`relu` as `piecewise`/`max`, `softplus` as `ln(1 + exp(x))`, `gelu`/`silu`/`mish`/`elu`/`selu` via `exp`/`tanh`; `softmax` needs the sum of the outputs, which is an assignment rule of its own); convolution, pooling and normalization layers inside the RHS are out of scope and are reported as gaps (see below).

Then the network parameters are `FitParameter`s (one per array element, on the linear scale, unbounded unless the parameter table says otherwise), the frozen ones stay changes, the objective and the optimizer are untouched, and the gradient of the test suite is compared against the finite difference jacobian `sbmlsim.fit.fisher` already builds, or against the forward sensitivities of roadrunner if they are exposed.

Open points to decide during the implementation:

- The likelihood: the test suite compares `llh` of the noise model (`noiseFormula`, `noiseDistribution`) and not a least squares cost. The reader currently maps a numeric `noiseFormula` to a weight; a `log_likelihood` of the resolved problem is needed for the comparison (this is a gap of the PEtab v2 layer in general, not only of SciML).
- Priors on the NN parameters (032–034) depend on #190.
- Whether the compiled SBML model is written next to the problem (like `observables.py` does) or kept in memory.
- Whether the `petab_sciml` package becomes a dependency (it pulls `mkstd`, `ruamel.yaml` and `h5py`, `torch` is optional) or the NN YAML and the HDF5 layout are read directly with `pydantic` and `h5py`. `h5py` is currently not a dependency (`examples/comparison` imports it optionally).
- Dropout, batch and instance normalization have a training mode which a deterministic ODE fit does not have; the test suite averages dropout over forward passes. Evaluation mode only.

## Tasks

- [ ] `sbmlsim.fit.petab_v2.sciml` (or `sbmlsim.sciml`): pydantic models of the NN YAML (`layers`, `forward`, inputs) and of the `sciml` extension block, reading of the HDF5 array files (`inputs`, `parameters`, `metadata/pytorch_format`, condition specific inputs `cond1;cond2` and the global `0`), the hybridization table and the NN entries of the mapping table (`net.inputs[i][j]`, `net.outputs[i][j]`, `net.parameters[layer].array[idx]`), the parameter table rows of networks, layers and arrays with the precedence rule
- [ ] numpy forward pass of the NN YAML: `Linear`, `Bilinear`, `Flatten`, the activations and the tensor ops of `petab_sciml.constants.TensorOps`; the convolution, pooling, normalization and dropout layers in evaluation mode as far as they are needed for `ml_model_import`, with the row major/PyTorch axis convention of the arrays
- [ ] compilation of a feed-forward network into SBML assignment rules and parameters for the RHS pattern, activations as MathML, unsupported layers raise
- [ ] `PetabReader`: the three patterns, several networks per problem, frozen networks/layers/arrays, inputs which are estimated parameters or condition parameters or arrays, expressions as inputs (035), two input arguments (029), several outputs (019–022)
- [ ] `PetabExporter`: write a problem with networks back (the extension block, the tables, the arrays), so a round trip is exact; the `sbmlsim` extension keeps the settings and the kinds as today
- [ ] a log-likelihood of the resolved problem for the comparison with `llh` of the test suite
- [ ] `gaps.py`: the SciML gaps (`format: pytorch | equinox | lux.jl`, layers which cannot be compiled into SBML, training mode of dropout and normalization, `array` inputs which differ per condition)
- [ ] `tests/fit/test_petab_sciml.py`: run the test suite (`ml_model_import`, `sciml_problem_import`, `initialization`) against a checkout of `petab_sciml_testsuite`, e.g. as a submodule under `tests/data/` or downloaded into a cache like the SBML test suite of #201, with the per case tolerances of `solutions.yaml`; unsupported cases are skipped with the reason
- [ ] `examples/sciml/`: the neural ODE of the getting started tutorial and the Lotka-Volterra problem of case 001 fitted with `sbmlsim`, including the report
- [ ] documentation: `docs/petab.md` section on SciML, supported layers table (the same table as https://petab-sciml.readthedocs.io/latest/layers.html with a column for `sbmlsim`), gaps, and the entry in `docs/references.md`; request the `sbmlsim` column in the layers table of the PEtab SciML documentation once the support lands

## References

- Paper: https://arxiv.org/abs/2608.20184 (Persson et al. 2026, PEtab SciML)
- Format and library: https://github.com/PEtab-dev/petab_sciml, https://petab-sciml.readthedocs.io
- Test suite: https://github.com/PEtab-dev/petab_sciml_testsuite
- Benchmark collection: https://github.com/sebapersson/Benchmark-Models-PEtab-SciML
- PEtab.jl: https://github.com/sebapersson/PEtab.jl (Persson et al. 2025, Bioinformatics 41:btaf497)
- AMICI JAX: https://github.com/AMICI-dev/AMICI/tree/master/python/sdist/amici/jax (Fröhlich et al. 2021, Bioinformatics 37:3676)
- PEtab v2: Pathirana et al. 2026, J Integr Bioinform, doi:10.1515/jib-2026-0004
- Related: #202 (PEtab v2), #190 (priors), #201 (SBML test suite)
